src/auto-evo/simulation/PopulationSimulation.cs

```diff
@@ -117,36 +117,132 @@ private static void RunSimulationStep(SimulationConfiguration parameters, List<S
+
+            // Calculate the total scores of each type in the current patch
+            foreach (MicrobeSpecies currentMicrobeSpecies in species)
+            {
+                totalPhotosynthesisScore += GetPhotosynthesisScore(currentMicrobeSpecies);
+                totalChemosynthesisScore += GetChemosynthesisScore(currentMicrobeSpecies);
+                totalPredationScore += GetPredationScore(currentMicrobeSpecies);
+            }
+
+            // Avoid division by 0
+            totalPhotosynthesisScore = Math.Max(MathUtils.EPSILON, totalPhotosynthesisScore);
+            totalChemosynthesisScore = Math.Max(MathUtils.EPSILON, totalChemosynthesisScore);
+            totalPredationScore = Math.Max(MathUtils.EPSILON, totalPredationScore);
+
+            // Calculate the share of environmental energy captured by each species
+            var energyAvailableForPredation = 0.0f;
+
+            foreach (MicrobeSpecies currentMicrobeSpecies in species)
+            {
+                var currentSpeciesEnergy = 0.0f;
+
+                currentSpeciesEnergy += sunlightInPatch
+                    * GetPhotosynthesisScore(currentMicrobeSpecies) / totalPhotosynthesisScore;
+
+                currentSpeciesEnergy += hydrogenSulfideInPatch
+                    * GetChemosynthesisScore(currentMicrobeSpecies) / totalChemosynthesisScore;
+
+                energyAvailableForPredation += 0.5f * currentSpeciesEnergy;
+                speciesEnergies.Add(currentMicrobeSpecies, currentSpeciesEnergy);
+            }
+
+            // Calculate the share of predation done by each species
+            // Then update populations
+            foreach (MicrobeSpecies currentMicrobeSpecies in species)
+            {
+                speciesEnergies[currentMicrobeSpecies] += energyAvailableForPredation
+                    * GetPredationScore(currentMicrobeSpecies) / totalPredationScore;
+
+                var newPopulation = (int)(speciesEnergies[currentMicrobeSpecies]
+                    / Math.Pow(currentMicrobeSpecies.Organelles.Count(), 1.3f));
+
+                populations.AddPopulationResultForSpecies(currentMicrobeSpecies, patch, newPopulation);
+            }
+        }
+
+        private static float GetPhotosynthesisScore(MicrobeSpecies species)
+        {
+            var photosynthesisScore = 0.0f;
+
+            var sunlight = SimulationParameters.Instance.GetCompound("sunlight");
+            var glucose = SimulationParameters.Instance.GetCompound("glucose");
 
-            foreach (var currentSpecies in species)
+            foreach (var organelle in species.Organelles)
             {
-                int currentPopulation = populations.GetPopulationInPatch(currentSpecies, patch);
-                int populationChange = random.Next(
-                    -Constants.AUTO_EVO_RANDOM_POPULATION_CHANGE, Constants.AUTO_EVO_RANDOM_POPULATION_CHANGE + 1);
+                var processesDoneByOrganelle = organelle.Definition.RunnableProcesses;
```

Comment:
This local is not needed, right? The foreach loop could just directly use `organelle.Definition.RunnableProcesses`